File: frontend/features/loop/catalog.ts
import { CardKind, LoopStage, WorkflowNode } from "@/lib/api/generated/model";

export const LOOP_STAGE_CATALOG = [
  {
    id: LoopStage.grilling,
    name: "Grilling",
    description: "Clarify the research idea through questions you confirm.",
    nodes: [WorkflowNode.idea_interpretation, WorkflowNode.idea_decomposition],
  },
  {
    id: LoopStage.related_work,
    name: "Related work",
    description: "Locate and assess prior work.",
    nodes: [WorkflowNode.research_inputs, WorkflowNode.related_work],
  },
  {
    id: LoopStage.gap,
    name: "Gap",
    description: "Synthesize the research gap from related work.",
    nodes: [WorkflowNode.gap],
  },
  {
    id: LoopStage.contribution,
    name: "Contribution",
    description: "Choose a contribution direction after the gap.",
    nodes: [WorkflowNode.contribution],
  },
  {
    id: LoopStage.claims_evidence,
    name: "Claims/evidence",
    description: "State claims and the evidence that would support them.",
    nodes: [WorkflowNode.claims, WorkflowNode.evidence],
  },
  {
    id: LoopStage.experiment_planning,
    name: "Experiment planning",
    description: "Plan tests that could confirm or refute the claims.",
    nodes: [WorkflowNode.experiment_plan, WorkflowNode.feasibility],
  },
  {
    id: LoopStage.spec_draft,
    name: "Spec Draft",
    description: "Read the Produced Spec Version before Independent judges.",
    nodes: [],
  },
  {
    id: LoopStage.independent_judges,
    name: "Independent judges",
    description: "Separate judgement of the Research Spec.",
    nodes: [
      WorkflowNode.gap_judge,
      WorkflowNode.contribution_judge,
      WorkflowNode.evidence_judge,
      WorkflowNode.experiment_judge,
      WorkflowNode.conference_judge,
      WorkflowNode.aggregator,
    ],
  },
  {
    id: LoopStage.readiness,
    name: "Readiness",
    description: "Evaluate readiness criteria. Not conference acceptance.",
    nodes: [],
  },
] as const;

type CatalogStageId = (typeof LOOP_STAGE_CATALOG)[number]["id"];
type MissingStage = Exclude<LoopStage, CatalogStageId>;
type ExtraStage = Exclude<CatalogStageId, LoopStage>;
const _allStagesPresent: MissingStage extends never ? true : MissingStage = true;
const _noExtraStages: ExtraStage extends never ? true : ExtraStage = true;
void _allStagesPresent;
void _noExtraStages;

type CatalogNode = (typeof LOOP_STAGE_CATALOG)[number]["nodes"][number];
type MissingNode = Exclude<WorkflowNode, CatalogNode>;
type ExtraNode = Exclude<CatalogNode, WorkflowNode>;
const _allNodesPresent: MissingNode extends never ? true : MissingNode = true;
const _noExtraNodes: ExtraNode extends never ? true : ExtraNode = true;
void _allNodesPresent;
void _noExtraNodes;

const INVALIDATION_CHILDREN: Record<WorkflowNode, readonly WorkflowNode[]> = {
  [WorkflowNode.idea_interpretation]: [WorkflowNode.idea_decomposition],
  [WorkflowNode.idea_decomposition]: [WorkflowNode.research_inputs, WorkflowNode.gap_judge],
  [WorkflowNode.research_inputs]: [WorkflowNode.related_work],
  [WorkflowNode.related_work]: [
    WorkflowNode.gap,
    WorkflowNode.gap_judge,
    WorkflowNode.contribution_judge,
    WorkflowNode.evidence_judge,
  ],
  [WorkflowNode.gap]: [
    WorkflowNode.contribution,
    WorkflowNode.gap_judge,
    WorkflowNode.contribution_judge,
  ],
  [WorkflowNode.contribution]: [
    WorkflowNode.claims,
    WorkflowNode.contribution_judge,
    WorkflowNode.experiment_judge,
  ],
  [WorkflowNode.claims]: [
    WorkflowNode.evidence,
    WorkflowNode.evidence_judge,
    WorkflowNode.experiment_judge,
  ],
  [WorkflowNode.evidence]: [
    WorkflowNode.experiment_plan,
    WorkflowNode.gap_judge,
    WorkflowNode.contribution_judge,
  ],
  [WorkflowNode.experiment_plan]: [WorkflowNode.feasibility, WorkflowNode.experiment_judge],
  [WorkflowNode.feasibility]: [WorkflowNode.experiment_judge, WorkflowNode.conference_judge],
  [WorkflowNode.gap_judge]: [WorkflowNode.aggregator],
  [WorkflowNode.contribution_judge]: [WorkflowNode.aggregator],
  [WorkflowNode.evidence_judge]: [WorkflowNode.aggregator],
  [WorkflowNode.experiment_judge]: [WorkflowNode.aggregator],
  [WorkflowNode.conference_judge]: [WorkflowNode.aggregator],
  [WorkflowNode.aggregator]: [],
};

export const HANDLING_OPTION_TARGETS = [
  WorkflowNode.gap,
  WorkflowNode.contribution,
  WorkflowNode.claims,
  WorkflowNode.evidence,
  WorkflowNode.experiment_plan,
  WorkflowNode.idea_decomposition,
] as const;

export const WORKFLOW_NODE_LABELS: Record<WorkflowNode, string> = {
  [WorkflowNode.idea_interpretation]: "Idea interpretation",
  [WorkflowNode.idea_decomposition]: "Idea decomposition",
  [WorkflowNode.research_inputs]: "Research inputs",
  [WorkflowNode.related_work]: "Related work",
  [WorkflowNode.gap]: "Gap",
  [WorkflowNode.contribution]: "Contribution direction",
  [WorkflowNode.claims]: "Claims",
  [WorkflowNode.evidence]: "Evidence",
  [WorkflowNode.experiment_plan]: "Experiment plan",
  [WorkflowNode.feasibility]: "Feasibility",
  [WorkflowNode.gap_judge]: "Gap Judge",
  [WorkflowNode.contribution_judge]: "Contribution Judge",
  [WorkflowNode.evidence_judge]: "Evidence Judge",
  [WorkflowNode.experiment_judge]: "Experiment Judge",
  [WorkflowNode.conference_judge]: "Conference Judge",
  [WorkflowNode.aggregator]: "Aggregator",
};

export function isIndependentJudgeNode(node: WorkflowNode): boolean {
  return (
    stageForWorkflowNode(node) === LoopStage.independent_judges &&
    node !== WorkflowNode.aggregator
  );
}

export function stagePathNodes(stage: LoopStage): readonly WorkflowNode[] {
  if (stage === LoopStage.independent_judges) {
    return [];
  }
  return catalogStage(stage).nodes;
}

export function catalogStage(id: LoopStage) {
  const stage = LOOP_STAGE_CATALOG.find((entry) => entry.id === id);
  if (!stage) {
    throw new Error(`Loop Stage ${id} is missing from the catalog`);
  }
  return stage;
}

export function isLoopStage(value: string | null): value is LoopStage {
  return Object.values(LoopStage).some((stage) => stage === value);
}

export function stageForWorkflowNode(node: WorkflowNode): LoopStage {
  const stage = LOOP_STAGE_CATALOG.find((entry) =>
    (entry.nodes as readonly WorkflowNode[]).includes(node),
  );
  if (!stage) {
    throw new Error(`Workflow Node ${node} is missing from the Loop Stage catalog`);
  }
  return stage.id;
}

export const CARD_KIND_OWNERS: Record<CardKind, WorkflowNode[]> = {
  [CardKind.problem]: [WorkflowNode.idea_decomposition],
  [CardKind.research_question]: [WorkflowNode.idea_decomposition],
  [CardKind.constraint]: [WorkflowNode.idea_decomposition],
  [CardKind.open_question]: [WorkflowNode.idea_decomposition],
  [CardKind.gap]: [WorkflowNode.gap],
  [CardKind.contribution]: [WorkflowNode.contribution],
  [CardKind.claim]: [WorkflowNode.claims, WorkflowNode.evidence],
  [CardKind.evidence]: [WorkflowNode.evidence],
};

export const CARD_KIND_LABELS: Record<CardKind, string> = {
  [CardKind.problem]: "Problem",
  [CardKind.research_question]: "Research question",
  [CardKind.gap]: "Gap",
  [CardKind.contribution]: "Contribution",
  [CardKind.claim]: "Claim",
  [CardKind.evidence]: "Evidence",
  [CardKind.constraint]: "Constraint",
  [CardKind.open_question]: "Open question",
};

export function ownedCardKinds(node: WorkflowNode): CardKind[] {
  return (Object.values(CardKind) as CardKind[]).filter((kind) => CARD_KIND_OWNERS[kind].includes(node));
}

export function resolveSelectedStage(
  stageQuery: string | null,
  workingDraftNode: WorkflowNode,
): LoopStage {
  if (isLoopStage(stageQuery)) {
    return stageQuery;
  }
  return stageForWorkflowNode(workingDraftNode);
}

export function isWorkflowNode(value: string | null): value is WorkflowNode {
  return Object.values(WorkflowNode).some((node) => node === value);
}

export type NavStop = {
  stage: LoopStage;
  node?: WorkflowNode;
  exportScratch?: boolean;
  specVersionId?: string;
};

export function navStops(): NavStop[] {
  return LOOP_STAGE_CATALOG.flatMap((stage) => {
    if (stage.id === LoopStage.independent_judges || stage.nodes.length === 0) {
      return [{ stage: stage.id }];
    }
    return [...stage.nodes].map((node) => ({ stage: stage.id, node }));
  });
}

export function sessionHref(sessionId: string, stop: NavStop): string {
  const params = new URLSearchParams({ stage: stop.stage });
  if (stop.node) {
    params.set("node", stop.node);
  }
  if (stop.exportScratch && stop.stage === LoopStage.readiness) {
    params.set("export_scratch", "1");
    if (stop.specVersionId) {
      params.set("spec_version", stop.specVersionId);
    }
  }
  return `/sessions/${sessionId}?${params.toString()}`;
}

export function isExportScratchEditorOpen(
  stage: LoopStage | null,
  searchParams: { get: (name: string) => string | null },
): boolean {
  return stage === LoopStage.readiness && searchParams.get("export_scratch") === "1";
}

export function railStop(stage: LoopStage): NavStop {
  if (stage === LoopStage.independent_judges) {
    return { stage };
  }
  const nodes = catalogStage(stage).nodes as readonly WorkflowNode[];
  return nodes[0] ? { stage, node: nodes[0] } : { stage };
}

export function resolveSelectedNode(
  stage: LoopStage,
  nodeQuery: string | null,
  workingDraftNode: WorkflowNode,
): WorkflowNode | undefined {
  const nodes = catalogStage(stage).nodes as readonly WorkflowNode[];
  if (stage === LoopStage.independent_judges) {
    if (nodes.includes(workingDraftNode)) {
      return workingDraftNode;
    }
    return WorkflowNode.aggregator;
  }
  if (nodes.length === 0) {
    return undefined;
  }
  if (isWorkflowNode(nodeQuery) && nodes.includes(nodeQuery)) {
    return nodeQuery;
  }
  if (!nodeQuery && nodes.includes(workingDraftNode)) {
    return workingDraftNode;
  }
  return nodes[0];
}

export function adjacentStop(current: NavStop, delta: -1 | 1): NavStop | null {
  const stops = navStops();
  const index = stops.findIndex(
    (stop) => stop.stage === current.stage && stop.node === current.node,
  );
  if (index < 0) {
    return null;
  }
  return stops[index + delta] ?? null;
}

export function workingDraftStop(node: WorkflowNode): NavStop {
  const stage = stageForWorkflowNode(node);
  if (stage === LoopStage.independent_judges) {
    return { stage };
  }
  return { stage, node };
}

export function ancestors(node: WorkflowNode): Set<WorkflowNode> {
  const found = new Set<WorkflowNode>();
  const stack = Object.entries(INVALIDATION_CHILDREN)
    .filter(([, children]) => children.includes(node))
    .map(([parent]) => parent as WorkflowNode);
  while (stack.length > 0) {
    const parent = stack.pop();
    if (!parent || found.has(parent)) {
      continue;
    }
    found.add(parent);
    stack.push(
      ...Object.entries(INVALIDATION_CHILDREN)
        .filter(([, children]) => children.includes(parent))
        .map(([grandparent]) => grandparent as WorkflowNode),
    );
  }
  return found;
}

export function descendants(node: WorkflowNode): Set<WorkflowNode> {
  const found = new Set<WorkflowNode>();
  const stack = [...INVALIDATION_CHILDREN[node]];
  while (stack.length > 0) {
    const child = stack.pop();
    if (!child || found.has(child)) {
      continue;
    }
    found.add(child);
    stack.push(...INVALIDATION_CHILDREN[child]);
  }
  return found;
}

export function upstreamOfStage(stage: LoopStage): Set<WorkflowNode> {
  const stageNodes = new Set<WorkflowNode>(catalogStage(stage).nodes);
  const upstream = new Set<WorkflowNode>();
  for (const node of stageNodes) {
    for (const parent of ancestors(node)) {
      if (!stageNodes.has(parent)) {
        upstream.add(parent);
      }
    }
  }
  return upstream;
}

File: backend/app/modules/loop/catalog.py
"""Loop catalog: Workflow Nodes, Loop Stages, invalidation edges, Card owners."""

from enum import StrEnum


class WorkflowNode(StrEnum):
    IDEA_INTERPRETATION = "idea_interpretation"
    IDEA_DECOMPOSITION = "idea_decomposition"
    RESEARCH_INPUTS = "research_inputs"
    RELATED_WORK = "related_work"
    GAP = "gap"
    CONTRIBUTION = "contribution"
    CLAIMS = "claims"
    EVIDENCE = "evidence"
    EXPERIMENT_PLAN = "experiment_plan"
    FEASIBILITY = "feasibility"
    GAP_JUDGE = "gap_judge"
    CONTRIBUTION_JUDGE = "contribution_judge"
    EVIDENCE_JUDGE = "evidence_judge"
    EXPERIMENT_JUDGE = "experiment_judge"
    CONFERENCE_JUDGE = "conference_judge"
    AGGREGATOR = "aggregator"


class LoopStage(StrEnum):
    GRILLING = "grilling"
    RELATED_WORK = "related_work"
    GAP = "gap"
    CONTRIBUTION = "contribution"
    CLAIMS_EVIDENCE = "claims_evidence"
    EXPERIMENT_PLANNING = "experiment_planning"
    SPEC_DRAFT = "spec_draft"
    INDEPENDENT_JUDGES = "independent_judges"
    READINESS = "readiness"


class CardKind(StrEnum):
    PROBLEM = "problem"
    RESEARCH_QUESTION = "research_question"
    GAP = "gap"
    CONTRIBUTION = "contribution"
    CLAIM = "claim"
    EVIDENCE = "evidence"
    CONSTRAINT = "constraint"
    OPEN_QUESTION = "open_question"


class NodeHeadStatus(StrEnum):
    EMPTY = "empty"
    CURRENT = "current"
    STALE = "stale"


class DecisionKind(StrEnum):
    CONFIRM = "confirm"
    EDIT = "edit"
    PICK = "pick"
    REVERT = "revert"
    EXPORT_ACK = "export_ack"


WORKFLOW_NODES: tuple[WorkflowNode, ...] = tuple(
    node for node in WorkflowNode if node is not WorkflowNode.EVIDENCE
)

LOOP_STAGE_NODES: dict[LoopStage, tuple[WorkflowNode, ...]] = {
    LoopStage.GRILLING: (
        WorkflowNode.IDEA_INTERPRETATION,
        WorkflowNode.IDEA_DECOMPOSITION,
    ),
    LoopStage.RELATED_WORK: (
        WorkflowNode.RESEARCH_INPUTS,
        WorkflowNode.RELATED_WORK,
    ),
    LoopStage.GAP: (WorkflowNode.GAP,),
    LoopStage.CONTRIBUTION: (WorkflowNode.CONTRIBUTION,),
    LoopStage.CLAIMS_EVIDENCE: (WorkflowNode.CLAIMS,),
    LoopStage.EXPERIMENT_PLANNING: (
        WorkflowNode.EXPERIMENT_PLAN,
        WorkflowNode.FEASIBILITY,
    ),
    LoopStage.SPEC_DRAFT: (),
    LoopStage.INDEPENDENT_JUDGES: (
        WorkflowNode.GAP_JUDGE,
        WorkflowNode.CONTRIBUTION_JUDGE,
        WorkflowNode.EVIDENCE_JUDGE,
        WorkflowNode.EXPERIMENT_JUDGE,
        WorkflowNode.CONFERENCE_JUDGE,
        WorkflowNode.AGGREGATOR,
    ),
    LoopStage.READINESS: (),
}

# parent → children (invalidation DAG). Spec Version is not a node; feasibility feeds conference_judge.
_EDGES: dict[WorkflowNode, tuple[WorkflowNode, ...]] = {
    WorkflowNode.IDEA_INTERPRETATION: (WorkflowNode.IDEA_DECOMPOSITION,),
    WorkflowNode.IDEA_DECOMPOSITION: (
        WorkflowNode.RESEARCH_INPUTS,
        WorkflowNode.GAP_JUDGE,
    ),
    WorkflowNode.RESEARCH_INPUTS: (WorkflowNode.RELATED_WORK,),
    WorkflowNode.RELATED_WORK: (
        WorkflowNode.GAP,
        WorkflowNode.GAP_JUDGE,
        WorkflowNode.CONTRIBUTION_JUDGE,
        WorkflowNode.EVIDENCE_JUDGE,
    ),
    WorkflowNode.GAP: (
        WorkflowNode.CONTRIBUTION,
        WorkflowNode.GAP_JUDGE,
        WorkflowNode.CONTRIBUTION_JUDGE,
    ),
    WorkflowNode.CONTRIBUTION: (
        WorkflowNode.CLAIMS,
        WorkflowNode.CONTRIBUTION_JUDGE,
        WorkflowNode.EXPERIMENT_JUDGE,
    ),
    WorkflowNode.CLAIMS: (
        WorkflowNode.EXPERIMENT_PLAN,
        WorkflowNode.EVIDENCE_JUDGE,
        WorkflowNode.EXPERIMENT_JUDGE,
    ),
    WorkflowNode.EVIDENCE: (),
    WorkflowNode.EXPERIMENT_PLAN: (
        WorkflowNode.FEASIBILITY,
        WorkflowNode.EXPERIMENT_JUDGE,
    ),
    WorkflowNode.FEASIBILITY: (
        WorkflowNode.EXPERIMENT_JUDGE,
        WorkflowNode.CONFERENCE_JUDGE,
    ),
    WorkflowNode.GAP_JUDGE: (WorkflowNode.AGGREGATOR,),
    WorkflowNode.CONTRIBUTION_JUDGE: (WorkflowNode.AGGREGATOR,),
    WorkflowNode.EVIDENCE_JUDGE: (WorkflowNode.AGGREGATOR,),
    WorkflowNode.EXPERIMENT_JUDGE: (WorkflowNode.AGGREGATOR,),
    WorkflowNode.CONFERENCE_JUDGE: (WorkflowNode.AGGREGATOR,),
    WorkflowNode.AGGREGATOR: (),
}

FIVE_JUDGE_NODES: tuple[WorkflowNode, ...] = (
    WorkflowNode.GAP_JUDGE,
    WorkflowNode.CONTRIBUTION_JUDGE,
    WorkflowNode.EVIDENCE_JUDGE,
    WorkflowNode.EXPERIMENT_JUDGE,
    WorkflowNode.CONFERENCE_JUDGE,
)

CARD_KIND_OWNERS: dict[CardKind, tuple[WorkflowNode, ...]] = {
    CardKind.PROBLEM: (WorkflowNode.IDEA_DECOMPOSITION,),
    CardKind.RESEARCH_QUESTION: (WorkflowNode.IDEA_DECOMPOSITION,),
    CardKind.CONSTRAINT: (WorkflowNode.IDEA_DECOMPOSITION,),
    CardKind.OPEN_QUESTION: (WorkflowNode.IDEA_DECOMPOSITION,),
    CardKind.GAP: (WorkflowNode.GAP,),
    CardKind.CONTRIBUTION: (WorkflowNode.CONTRIBUTION,),
    CardKind.CLAIM: (WorkflowNode.CLAIMS,),
    CardKind.EVIDENCE: (WorkflowNode.CLAIMS,),
}

_OWNER_KINDS: dict[WorkflowNode, tuple[CardKind, ...]] = {}
for _kind, _owners in CARD_KIND_OWNERS.items():
    for _owner in _owners:
        _OWNER_KINDS[_owner] = (*_OWNER_KINDS.get(_owner, ()), _kind)


def owned_kinds(node: WorkflowNode) -> tuple[CardKind, ...]:
    return _OWNER_KINDS.get(node, ())


def active_workflow_node(node: WorkflowNode) -> WorkflowNode:
    if node is WorkflowNode.EVIDENCE:
        return WorkflowNode.CLAIMS
    return node


def _card_body_nonblank(body: object) -> bool:
    if not isinstance(body, dict):
        return False
    for key in ("text", "statement", "claim", "evidence"):
        value = body.get(key)
        if isinstance(value, str) and value.strip():
            return True
    return False


def claims_confirmable(cards: object) -> bool:
    kinds: set[CardKind] = set()
    if not isinstance(cards, list):
        return False
    for card in cards:
        kind_raw = getattr(card, "kind", None)
        body = getattr(card, "body", None)
        if kind_raw is None and isinstance(card, dict):
            kind_raw = card.get("kind")
            body = card.get("body")
        try:
            kind = kind_raw if isinstance(kind_raw, CardKind) else CardKind(str(kind_raw))
        except ValueError:
            continue
        if kind in (CardKind.CLAIM, CardKind.EVIDENCE) and _card_body_nonblank(body):
            kinds.add(kind)
    return CardKind.CLAIM in kinds and CardKind.EVIDENCE in kinds


HANDLING_OPTION_TARGETS: frozenset[str] = frozenset(
    {
        WorkflowNode.GAP.value,
        WorkflowNode.CONTRIBUTION.value,
        WorkflowNode.CLAIMS.value,
        WorkflowNode.EXPERIMENT_PLAN.value,
        WorkflowNode.IDEA_DECOMPOSITION.value,
    }
)


def descendants(node: WorkflowNode) -> frozenset[WorkflowNode]:
    found: set[WorkflowNode] = set()
    stack = list(_EDGES[node])
    while stack:
        child = stack.pop()
        if child in found:
            continue
        found.add(child)
        stack.extend(_EDGES[child])
    return frozenset(found)


def ancestors(node: WorkflowNode) -> frozenset[WorkflowNode]:
    found: set[WorkflowNode] = set()
    stack = [parent for parent, children in _EDGES.items() if node in children]
    while stack:
        parent = stack.pop()
        if parent in found:
            continue
        found.add(parent)
        stack.extend(p for p, children in _EDGES.items() if parent in children)
    return frozenset(found)


def upstream_of_stage(stage: LoopStage) -> frozenset[WorkflowNode]:
    stage_nodes = set(LOOP_STAGE_NODES[stage])
    up: set[WorkflowNode] = set()
    for node in stage_nodes:
        up.update(ancestors(node) - stage_nodes)
    return frozenset(up)


def first_needs_work(
    stage: LoopStage,
    status_by_node: dict[WorkflowNode, NodeHeadStatus],
) -> WorkflowNode | None:
    for node in LOOP_STAGE_NODES[stage]:
        if status_by_node[node] in (NodeHeadStatus.EMPTY, NodeHeadStatus.STALE):
            return node
    return None


def prepare_landing(
    stage: LoopStage,
    status_by_node: dict[WorkflowNode, NodeHeadStatus],
) -> WorkflowNode | None:
    needs_work = first_needs_work(stage, status_by_node)
    if needs_work is None:
        return None
    if stage is LoopStage.INDEPENDENT_JUDGES:
        return WorkflowNode.AGGREGATOR
    return needs_work
